### PR TITLE
Add bash script to check driver on some distros.

### PR DESCRIPTION
**Pull request by aquadzn**

### CODE REVIEW

File: scripts/check_driver.sh [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        then
+            sudo apt-get install mesa-va-drivers -y
+        elif [[ $pilot == *"NVIDIA"* ]]
+        then
+            echo -e "For NVIDIA users, you have 2 options:\n"
+            echo -e "\t* Use the Nouveau driver and install the old NVIDIA firmware that provides support for the VA API."
+            echo -e "\t* Use the Arekinath patch to provides VA API compatibility for recent NVIDIA cards."
+        fi
+
+        if $(vainfo |& grep -q -e "H265" -e "HEVC")
```
Comment (NicolasGuilloux) [inline scripts/check_driver.sh]: It would be great to all of the checks inside a function to avoid boiler plate

**Comment by aquadzn** [comment]
done 👍

**Review by agentcobra — APPROVED** [review]
look OK for me

**Comment by NicolasGuilloux** [comment]
There is some stuff that I dislike in the script. It tries to install the driver evenif the vainfo which is very instrusive if we widespread this script for all community.

A better approach would be to test vainfo, and if it is not optimal, it checks if all the appropriate drivers are installed. If not, it asks to actually install them and recheck.

Moreover, a small warning about why we use `sudo` would be great aswell. Many users of the scripts we made are not good Linux users, and I guess they will start the command several times and enter their password several times aswell. Limiting that would be great :)

File: scripts/check_driver.sh [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    echo -e "\t* Use the Nouveau driver and install the old NVIDIA firmware that provides support for the VA API."
+    echo -e "\t* Use the Arekinath patch to provides VA API compatibility for recent NVIDIA cards."
+}
+
+DISTRO=$(cat /etc/*-release |& awk -F= '/^NAME/ { print $2 }' | sed 's/"//g')
+case $DISTRO in
+    "Ubuntu" | "Debian GNU/Linux")
+        if ! type "vainfo" > /dev/null
+        then
+            sudo apt-get update && sudo apt-get install vainfo libva-glx2
```
Comment (NicolasGuilloux) [inline scripts/check_driver.sh]: `sudo` warning (cf PR comment)

File: scripts/check_driver.sh [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            sudo sh -c "echo 'export VDPAU_DRIVER=va_gl' >> /etc/profile"
+        fi
+        if [[ $pilot == *"NVIDIA"* ]]
+        then
+            nvidia
+        fi
+
+        check_codec
+        ;;
+esac
```
Comment (NicolasGuilloux) [inline scripts/check_driver.sh]: No fallback for unrecognized OS ?

File: scripts/check_driver.sh [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        echo -e "- H264 found."
+    else
+        echo -e "\nH264 not found.\nPlease generate a report with the following command:\n\t- curl https://raw.githubusercontent.com/NicolasGuilloux/blade-shadow-beta/master/scripts/report.pl | perl"
+    fi
+}
+
+nvidia() {
+    echo -e "For NVIDIA users, you have 2 options:\n"
+    echo -e "\t* Use the Nouveau driver and install the old NVIDIA firmware that provides support for the VA API."
+    echo -e "\t* Use the Arekinath patch to provides VA API compatibility for recent NVIDIA cards."
```
Comment (NicolasGuilloux) [inline scripts/check_driver.sh:28]: A link to the Arekinath patch

File: scripts/check_driver.sh [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        echo -e "\n- H265 / HEVC found."
+    else
+        echo -e "\nH265 or HEVC not found."
+    fi
+
+    if $(vainfo |& grep -q "H264")
+    then
+        echo -e "- H264 found."
+    else
+        echo -e "\nH264 not found.\nPlease generate a report with the following command:\n\t- curl https://raw.githubusercontent.com/NicolasGuilloux/blade-shadow-beta/master/scripts/report.pl | perl"
```
Comment (NicolasGuilloux) [inline scripts/check_driver.sh]: `Sadly, not supporting at least H264 means not supporting Shadow at all. To get some help, please generate a full report with the following command, and share it with the Discord community`

File: scripts/check_driver.sh [hunk 6]
```diff
@@ -0,0 +1,98 @@
+#!/bin/bash
+# https://github.com/aquadzn
+# ```bash check_driver.sh```
+
+
+check_codec() {
+    if $(vainfo |& grep -q -e "H265" -e "HEVC")
```
Comment (NicolasGuilloux) [inline scripts/check_driver.sh]: I don't know if it is a sufficient condition to have H265|HEVC inside the vainfo to actually support the H265 codec. These are 2 regex I used to check H265 and H264:

- `/VAProfileH264Main.*:\t(VAEntrypoint\w*)/`
- `/VAProfile(HEVC|H265)Main.*:\t(VAEntrypoint\w*)/`

File: scripts/check_driver.sh [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        then
+            nvidia
+        fi
+
+        check_codec
+        ;;
+    "Fedora")
+        if ! type "vainfo" > /dev/null
+        then
+            sudo dnf install librtmp libva-intel-hybrid-driver libvdpau-va-gl
```
Comment (NicolasGuilloux) [inline scripts/check_driver.sh]: Does it work for AMD Radeon ?

File: scripts/check_driver.sh [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+
+check_h264() {
+    if $(vainfo |& grep -q -E "VAProfileH264Main")
+    then
+        true
+    else
+        echo -e "❌ \e[91mH264 not found.\e[0m\nPlease generate a report with the following command:\n\t- curl https://raw.githubusercontent.com/NicolasGuilloux/blade-shadow-beta/master/scripts/report.pl | perl"
+        false
+    fi
+}
```
Comment (aquadzn) [inline scripts/check_driver.sh]: I changed the grep pattern and wrapped both into two functions

File: scripts/check_driver.sh [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+
+
+print_nvidia() {
+    echo -e "For NVIDIA users, you have 2 options:\n"
+    echo -e "\t* Use the Nouveau driver and install the old NVIDIA firmware that provides support for the VA API."
+    echo -e "\t* Use the Arekinath patch to provides VA API compatibility for recent NVIDIA cards."
+    echo -e "\t* https://gitlab.com/aar642/libva-vdpau-driver"
+}
```
Comment (aquadzn) [inline scripts/check_driver.sh:30]: Added URL to Nvidia warning

File: scripts/check_driver.sh [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+
+print_distro() {
+    echo -e "✅ \e[32mFound $1.\e[0m"
+}
```
Comment (aquadzn) [inline scripts/check_driver.sh:35]: Added function to print the distro

File: scripts/check_driver.sh [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+print_vainfo_warning() {
+    echo -e "❌ \e[91mvainfo not found.\e[0m\n"
+    echo -e "⏸️  \e[93mNeed permission to install vainfo. Please re-run the script after installation.\e[0m\n"
+    sleep 2
+}
+
+print_driver_warning() {
+    echo -e "⏸️  \e[93mNeed permission to install driver(s).\e[0m\n"
+    sleep 2
+}
```
Comment (aquadzn) [inline scripts/check_driver.sh:47]: Added two warnings for vainfo and drivers installation

File: scripts/check_driver.sh [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+
+
+main() {
+    echo -e "\n⚠️ \e[93mWarning\033[0m\nThis script will ask for sudo permission in order to install missings packages and/or drivers.\n"
+    sleep 3
```
Comment (aquadzn) [inline scripts/check_driver.sh:52]: Added a warning about the use of sudo

File: scripts/check_driver.sh [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+                    echo -e "❌ \e[91mPilot not found.\e[0m\n"
+                fi
+                return
+
+            fi
+            ;;
+        *)
+            echo -e "❌ \e[91mDistribution not found. Please try again.\e[0m\n"
+            return
+            ;;
```
Comment (aquadzn) [inline scripts/check_driver.sh:225]: Added a case when distro is not in case choices

**Review by aquadzn — COMMENTED** [review]
Good | Missing vainfo | Distro not found
--- | --- | ---
good | vainfo | distro

@NicolasGuilloux @agentcobra 😃

File: scripts/check_driver.sh (hunk 7 above)
Comment (aquadzn) [reply scripts/check_driver.sh]: Can't find any infos about Radeon so I skipped it

File: scripts/check_driver.sh [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+                then
+                    echo -e "❌ \e[91mH265 or HEVC not found.\e[0m"
+                fi
+
+                pilot=$(vainfo |& grep 'vainfo: Driver version:')
+                print_driver_warning
+                if [[ $pilot == *"Intel"* ]]
+                then
+                    sudo dnf install libva-intel-hybrid-driver libvdpau-va-gl
+                    sudo sh -c "echo 'export LIBVA_DRIVER_NAME=i965' >> /etc/profile"
```
Comment (aquadzn) [inline scripts/check_driver.sh]: #55

File: scripts/check_driver.sh [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+
+                if [[ check_h264 ]] && [[ check_h265_hevc ]]
+                then
+                    echo -e "✅ \e[32mH264 found.\e[0m"
+                    echo -e "✅ \e[32mH265 / HEVC found.\e[0m\n"
+                    return
+                fi
+                if [[ ! check_h264 ]]
+                then
+                    echo -e "❌ \e[91mH264 not found.\e[0m\nPlease generate a report with the following command:\n\t- curl https://raw.githubusercontent.com/NicolasGuilloux/blade-shadow-beta/master/scripts/report.pl | perl"
```
Comment (NicolasGuilloux) [inline scripts/check_driver.sh:75]: You already print this warning from the `check_h264` function, don't you? :p

File: scripts/check_driver.sh [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
```
Comment (NicolasGuilloux) [inline scripts/check_driver.sh:118]: Same here, you print it from the function in the if

File: scripts/check_driver.sh (hunk 14 above)
Comment (NicolasGuilloux) [reply scripts/check_driver.sh]: I disagree: all Intel GPU does not support the `i965`, and the `iHD` driver may be better as it should give better performance.

**Review by NicolasGuilloux — CHANGES_REQUESTED** [review]

File: scripts/check_driver.sh [hunk 17]
```diff
@@ -17,7 +17,6 @@ check_h264() {
     then
         true
     else
-        echo -e "❌ \e[91mH264 not found.\e[0m\nPlease generate a report with the following command:\n\t- curl https://raw.githubusercontent.com/NicolasGuilloux/blade-shadow-beta/master/scripts/report.pl | perl"
```
Comment (aquadzn) [inline scripts/check_driver.sh]: I forgot to remove it from the function (it wasn't printing)

File: scripts/check_driver.sh (hunk 15 above)
Comment (aquadzn) [reply scripts/check_driver.sh:75]: https://github.com/NicolasGuilloux/blade-shadow-beta/pull/54#discussion_r418288125

File: scripts/check_driver.sh (hunk 14 above)
Comment (aquadzn) [reply scripts/check_driver.sh]: reverted https://github.com/aquadzn/blade-shadow-beta/commit/945c0f8bf0ef034dcdf4df102163e4f42819b931

**Pull request merged**